cybozu/reactor.hpp

```diff
@@ -37,42 +47,38 @@ class resource {
 
     // Close the file descriptor.
     virtual ~resource() {
-        ::close(m_fd);
+        if( ! m_closed ) {
+            ::close(m_fd);
+        }
```

Comment:
Not acquiring the lock here means that there is no direct happens-before relationship between `with_fd` and this destructor. So I think it's NOT obvious why `with_fd` and the destructor is ordered.

Reply:
OK. I'll add the following comment.

> `m_closed` is only accessed by the reactor thread, so using it does not require a lock.
> `m_fd` is a constant, so there is no need to synchronize memory to get its value.
> It is guaranteed that there are no other threads that use this resource when it is being destructed, so closing the file descriptor is safe.

Reply:
We need to describe that `m_fd` is not read by another thread because the use of closed fd is not safe. Your comment does not describe this point.

Reply:
I don't understand.
It is guaranteed as explained in the design doc as follows.
https://github.com/cybozu/yrmcds/blob/master/docs/design.md#strategy-to-reclaim-shared-sockets

> Once the reactor observes idle state of all worker threads, resources in the save list can be destructed safely.

So, it is guaranteed that, when a resource is destructed by the reactor thread, no other threads are using it. I can put the link here if it does make sense for you.

Reply:
It was obvious for me because destructing (freeing the memory of) a resource is possible only when no other threads have access to it. Otherwise, it is a bug.

Reply:
> I can put the link here if it does make sense for you.

Please add the link.
Deleting resources is the most difficult aspect of yrmcds, so it is not too much to explain.

Reply:
OK. will do.